src/azul/portal_service.py

```diff
@@ -0,0 +1,82 @@
+from copy import deepcopy
+from typing import (
+    Sequence,
+)
+import json
+import logging
+
+import boto3
+from botocore.exceptions import ClientError
+
+from azul import config
+from azul.plugin import Plugin
+from azul.types import JSON
+
+log = logging.getLogger(__name__)
+
+
+class PortalService:
+
+    def get_portal_integrations_db(self) -> Sequence[JSON]:
+        """
+        Retrieve portal DB from S3.
+
+        If no object is present at the location specified in azul.config, then
+        a default hardcoded DB is obtained from the current plugin, uploaded to
+        S3, and returned.
+        """
+
+        client = boto3.client('s3')
+        try:
+            retrieval = client.get_object(Bucket=config.portal_integrations_db_bucket,
+                                          Key=config.portal_integrations_db_object)
+        except client.exceptions.NoSuchKey:
+            plugin = Plugin.load()
+            db = self.stage_db(plugin.portal_integrations_db())
+            log.info('Portal integration DB not found in S3; using and uploading hardcoded DB from plugin.')
+            json_bytes = bytes(json.dumps(db), 'UTF-8')
```

Comment:
```suggestion
            json_bytes = json.dumps(db).encode()
```

I don't know why the bytes constructor requires the encoding argument.

Reply:
```python
with BytesIO() as b:
    with TextIOWrapper(b) as w:
        json.dump({'ü':'ber'}, w)
        w.flush()
        print(repr(b.getvalue()))
    
b'{"\\u00fc": "ber"}'
```

Should use less space since the unencoded string is never materialized. However, I'm not advocating that approach for this PR.

Reply:
It would be easy to to add that as a utility function in `strings.py` since you've already written it.

Reply:
More like `json.py` but we would first need make reasonably sure that it is indeed an space improvement and has no serious drawbacks.

Reply:
Fair. I'll leave it out for now.